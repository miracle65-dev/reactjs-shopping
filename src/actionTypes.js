/**
 * @namespace actionTypes
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * Redux action types
 */
export const CART_ADD = 'cart/ADD';
export const CART_REMOVE = 'cart/REMOVE';
export const CART_UPDATE = 'cart/UPDATE';
export const CART_EMPTY = 'cart/EMPTY';
