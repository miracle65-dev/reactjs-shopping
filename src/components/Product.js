/**
 * @flow weak
 * @module Product
 * @extends React.Component
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * React component - Product form with price
 *
 */
import React, { Component, PropTypes } from 'react';
import animateScroll from 'react-scroll/lib/mixins/animate-scroll';
import capitalize from 'underscore.string/capitalize';

const
  /**
   * @static propTypes
   * @memberof Product
   *
   * @prop {string} name - Name to display. Required
   * @prop {string} path - Path to product. Required
   * @prop {number} price - Price (value only). Required
   * @prop {string} imagePath - Path to main image. Required
   * @prop {string} currency - Price currency. Default is £
   * @prop {Object.<string, Array<(string | number)>>} options - Custom product properties.
   * Default is {}
   * @prop {string} iconAddProductClassName - ClassName for cart icon on add to button.
   * Default is 'icon-cart-plus'
   */
  propTypes = {
    name: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    imagePath: PropTypes.string.isRequired,
    currency: PropTypes.string,
    options: PropTypes.objectOf(
      PropTypes.arrayOf(
        PropTypes.oneOfType(
          [
            PropTypes.string,
            PropTypes.number,
          ]
        )
      )
    ),
    iconAddProductClassName: PropTypes.string,
  },
  /**
  * @static containerPropTypes
  * @memberof Product
  *
  * @prop {Function<string>} onAddProduct - Callback to call when
  * user want to add product in his cart
  * Example: onAddProduct('macbook-case', { quantity: 30, properties: { colour: 'red' } });
  * Required.
  * @prop {React Element} CheckoutButton - Button in the bottom of product.
  * Default is 'icon-cart-plus'
  */
  containerPropTypes = {
    onAddProduct: PropTypes.func.isRequired,
    CheckoutButton: PropTypes.element.isRequired,
  },
  defaultProps = {
    currency: '£',
    options: {},
    iconAddProductClassName: 'icon-cart-plus',
  };

function getAbsoluteOffsetTop({ offsetTop, offsetParent } : HTMLElement | Object = {}) : number {
  return offsetTop + (offsetParent && getAbsoluteOffsetTop(offsetParent));
}

export default class Product extends Component {

  static propTypes = { ...propTypes, ...containerPropTypes };
  static defaultProps = defaultProps;

  state = {
    quantity: 1,
  };

  handleQuantityValueChange = (
    { target: { value } } : { target : HTMLInputElement, }
  ) => {
    +value > -1 ? this.setState({ quantity: +value }) : 0;
  }

  handleSelectInputValueChange = (
    name : string,                 // product property name
    list : Array<string | number>, // product property options list
    { target: { value: listOptionName } } : { target : HTMLInputElement, }
  ) => {
    this.setState({ [name]: list.indexOf(listOptionName) });
  }

  /*
   * Return form-group consits of select input element
   * label value will be capitalized name
   * options - the list array
   */
  createSelectInputFromArray(name : string, list : Array<string | number>) : React$Element<any> {
    return (
      <div key={ name } className="form-group row">
        <label htmlFor={ name } className="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-form-label">
          { capitalize(name) + ':' }
        </label>
        <div className="col-xs-9 col-sm-9 col-md-9 col-lg-9">
          <select
            onChange={ this.handleSelectInputValueChange.bind(this, name, list) }
            className="form-control"
            value={ list[this.state[name] || 0] }
          >
            {
              /*
               * Generate select input options based on list values
               */
              list.map((v, k) =>
                <option key={ v + k } value={ v }>
                  { capitalize(v) }
                </option>
              )
            }
          </select>
        </div>
      </div>
    );
  }

  generateProductProps() : Object {
    const { options, price, name, imagePath, currency } = this.props;
    const { quantity } = this.state;
    return (
      {
        quantity,
        properties: Object.entries(options).reduce((obj, [propName, list]) =>
        ({
          ...obj,
          [propName]: list[this.state[propName] || 0],
        })
      , {}),
        productInfo: {
          name,
          price,
          currency,
          imagePath,
        },
      }
    );
  }

  addProductFormSubmit = (event : Event) => {
    event.preventDefault();
    const { path, onAddProduct } = this.props;
    const { quantity } = this.state;
    setTimeout(target =>
      animateScroll.scrollTo(
        getAbsoluteOffsetTop(
          target.children[target.children.length - 2]) - 5, {
            duration: 750,
            delay: 0,
            smooth: true,
          })
    , 50, event.target);
    return quantity
           &&
           onAddProduct(
             path,
             this.generateProductProps()
           );
  }

  render() {
    const { price, currency, options, CheckoutButton, iconAddProductClassName } = this.props;

    return (
      <div>
        <p>
          <strong>
            Price: { currency + price.toFixed(2) }
          </strong>
        </p>
        <form className="m-t-1" onSubmit={ this.addProductFormSubmit }>
          {
            Object.entries(options).map(
              option => this.createSelectInputFromArray(...option)
            )
          }
          <div className="form-group row">
            <label htmlFor="product-quantity" className="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-form-label">
              Quantity:
            </label>
            <div className="col-xs-9 col-sm-9 col-md-9 col-lg-9">
              <input onChange={ this.handleQuantityValueChange } className="form-control" type="number" value={ this.state.quantity } />
            </div>
          </div>
          <button type="submit" className="btn btn-success btn-block" disabled={ !this.state.quantity }>
            <i className={ iconAddProductClassName } />
            Add to cart
          </button>
          { CheckoutButton }
        </form>
      </div>
    );
  }
}
