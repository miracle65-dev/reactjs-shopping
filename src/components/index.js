export { default as Cart } from './Cart/Cart';
export { default as CheckoutButton } from './CheckoutButton/CheckoutButton';
export { default as Product } from './Product/Product';
export { default as CartProduct } from './Cart/CartProduct/CartProduct';
export { default as CartProductPropertyDescription } from
  './Cart/CartProduct/ProductPropertyDescription/ProductPropertyDescription';
export { default as ProductPropertyInput } from
  './Product/ProductPropertyInput/ProductPropertyInput';
