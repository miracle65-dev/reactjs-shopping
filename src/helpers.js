/**
 * @flow
 * @namespace helpers
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 */

import React from 'react';
import { animateScroll } from 'react-scroll';

import type {
  GenerateProductKey,
  DefaultLinkComponentProps,
} from './types';

/**
 * @typedef {Object} StyleConfig
 */
export type StyleConfig = {
  enabled?: boolean,
  duration?: number,
};

/** @ */
export type ReactStatelessComponent = (props: Object) => React$Element<*>;

/**
 * @memberof helpers
 */
export const generateStyle = (
  {
    enabled = true,
    duration = 1000,
    ...rest
  }: StyleConfig = {},
): Object =>
  enabled
    ? {
      WebkitAnimationDuration: `${duration / 1000}s`,
      animationDuration: `${duration / 1000}s`,
      WebkitAnimationFillMode: 'both',
      animationFillMode: 'both',
      ...rest,
    }
    : {};

/**
 * @memberof helpers
 */
export const animate = (options: StyleConfig | number): Object =>
  generateStyle(
    typeof options === 'object'
    ? options
    : { duration: options, },
  );

/**
 * @memberof helpers
 */
export const configure = (
  Component: Function,
  configuration: Object,
): ReactStatelessComponent => {
  const Configured: ReactStatelessComponent =
    props =>
      <Component {...configuration} {...props} />;
  Configured.displayName = `Configured(${
    Component.name || Component.displayName
  })`;
  return Configured;
};

/**
 * @memberof helpers
 */
export const isNaturalNumber = (num: number): boolean =>
  Number.isSafeInteger(num) && num > -1;

/**
 * @memberof helpers
 */
export const parseInteger = (num: string): number => {
  if (/^\d*$/.test(num))
    return Number.parseInt(num, 10) || 0;
  else return NaN;
};

/**
 * @memberof helpers
 */
export const getAbsoluteOffsetTop = (
  { offsetTop, offsetParent, }: HTMLElement = {},
): number =>
  offsetTop + (
    offsetParent
    && offsetParent instanceof HTMLElement
    && getAbsoluteOffsetTop(offsetParent)
  );

/*
 * key in format id/_property1-valueOfProperty1 etc
 */
export const generateProductKey: GenerateProductKey = (
  id,
  properties,
) =>
  Object
    .entries(properties)
    .reduce((acc: string, [ propName, propValue, ]) =>
      `${acc}_${propName}-${String(propValue)}`
    , `${id}/`);

/**
 * @memberof helpers
 */
export const DefaultLinkComponent = (
  { to, ...otherProps }: DefaultLinkComponentProps,
): React$Element<*> =>
  <a {...otherProps} href={to} />;

/**
 * @memberof helpers
 */
export const fixInputValueStartingWithZero = (
  target: HTMLInputElement, quantity: number,
) => {
  if (/^0+\d+$/.test(target.value))
    target.value = String(quantity);
};

/**
 * @memberof helpers
 */
export const scrollFunction = (
  target: EventTarget,
  scrollPosition: number | (currentTarget: Element) => number,
  scrollAnimationConfig: Object,
) => void (
  target instanceof Element
  && animateScroll.scrollTo(
    typeof scrollPosition === 'function'
    ? scrollPosition(target)
    : scrollPosition,
    scrollAnimationConfig,
  )
);
