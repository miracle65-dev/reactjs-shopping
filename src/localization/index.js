/**
 * @flow
 * @module localization
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * Default localization and API functions
 *
 */

import React from 'react';
import IntlMessageFormat from 'intl-messageformat';

import type {
  Localization,
  LocalizationPattern,
  MultiLocalization,
  GetLocalization,
} from '../types';

/**
 * @memberof localization
 */
export const defaultLocalization: MultiLocalization = {
  en: {
    cart: {
      shoppingCartTitle: {
        text: 'Shopping cart',
        component: 'h4',
      },
      productName: {
        text: '{localizedName}',
        component: 'h5',
      },
      quantityLabel: 'Quantity:',
      priceLabel: 'Price:',
      priceValue: '{localizedCurrency}{price, number}',
      totalLabel: 'Total:',
      totalValue: '{localizedCurrency}{total, plural, ' +
                  '=0 {0}' +
                  'other {#}}',
      remove: 'Remove',
      productPropertyLabel: '{localizedName}:',
      productPropertyValue: '{localizedValue}',
    },
    checkoutButton: {
      checkoutTotal:
        'Checkout (Grand total {localizedCurrency}{total, plural, ' +
        '=0 {0}' +
        'other {#}})',
    },
    product: {
      price: {
        text: 'Price: {localizedCurrency}{price}',
        component: 'strong',
      },
      quantityLabel: 'Quantity:',
      propertyLabel: '{localizedName}:',
      addToCart: 'Add to cart',
    },
  },
};

/**
 * @memberof localization
 */
export const getLocalization = (
  localization: Localization,
  language: string,
  id: string,
  params: Object = {},
): string | React$Element<*> => {
  const localizationPattern: LocalizationPattern = localization[id];

  if (typeof localizationPattern === 'undefined') {
    if (!process || process && process.env && process.env.NODE_ENV !== 'production') {
      console.error(
        `Missing localization for ${id}, language: ${language}`,
      );
    }
    return id;
  } else if (
    typeof localizationPattern === 'object'
    && typeof localizationPattern.text === 'string'
    && ['string', 'function'].includes(
      typeof localizationPattern.component,
    )
  ) return (
      React.createElement(
        localizationPattern.component,
        localizationPattern.props || {},
        new IntlMessageFormat(
          localizationPattern.text, language,
        ).format(params),
      )
  );
  else if (typeof localizationPattern === 'string')
    return (
      new IntlMessageFormat(localizationPattern, language).format(params)
    );
  else throw new Error(
    `Localization pattern error for ${id}, language: ${language}`,
  );
};

/**
 * @memberof localization
 */
export const getDefaultLocalization =
  (
    componentName: string,
    language: string = 'en',
    localization: Localization = {},
  ): GetLocalization =>
    (...args: Array<any>) =>
      getLocalization(
        {
          ...(
            defaultLocalization[language]
            ? defaultLocalization[language][componentName]
            : {}
          ),
          ...localization,
        },
        language,
        ...args,
      );
