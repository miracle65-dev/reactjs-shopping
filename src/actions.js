/**
 * @flow
 * @namespace actions
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * Redux action generators
 */
import * as actionTypes from './actionTypes';

/**
 * @memberof actions
 */
export const addToCart = (id : string, props : Object) : Object =>
  ({
    type: actionTypes.CART_ADD,
    id,
    ...props,
  });

/**
 * @memberof actions
 */
export const removeFromCart = (key : string) : Object =>
  ({
    type: actionTypes.CART_REMOVE,
    key,
  });

/**
 * @memberof actions
 */
export const updateCart = (key : string, updateProps : Object) : Object =>
  ({
    type: actionTypes.CART_UPDATE,
    key,
    updateProps,
  });

/**
 * @memberof actions
 */
export const emptyCart = () : Object =>
  ({
    type: actionTypes.CART_EMPTY,
  });
