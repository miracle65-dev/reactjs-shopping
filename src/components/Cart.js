/**
 * @flow weak
 * @module Cart
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * React stateless component which represents shopping cart
 */
import React, { PropTypes } from 'react';
import Link from 'react-router/lib/Link';
import ReactCSSTransitionGroup from 'react/lib/ReactCSSTransitionGroup';
import capitalize from 'underscore.string/capitalize';
import Transition from 'react-overlays/lib/Transition';

import { animate } from '../helpers';

const
  /**
   * @static propTypes
   * @memberof Cart
   *
   * @prop {Array<string>} productPropsToShow - Array of product's
   * props which need to be shown in cart.
   * Default is ['colour', 'size']
   *
   * @prop {boolean} showHeader - Show or hide header 'Shopping cart'.
   * Default is true
   * @prop {string} iconTrashClassName - ClassName for trash icon on remove button.
   * Default is 'icon-trash'
   */
  propTypes = {
    productPropsToShow: PropTypes.arrayOf(PropTypes.string),
    showHeader: PropTypes.bool,
    iconTrashClassName: PropTypes.string,
  },
  /**
   * @static containerPropTypes
   * @memberof Cart
   *
   * @prop {Object.<string, ProductType>} products - Products map. Required
   * @prop {Function<string, Object>} onUpdateProduct - Callback
   * function which will be called when product should be updated.
   * First arg is product's key in products, second - props to update.
   * For instance, it may be called like:
   * onUpdateProduct('/shop/macbook-case/_red', { quantity : 50});
   * Required.
   *
   * @prop {Function<string>} onRemoveProduct - Callback to call
   * when need to remove product from products. Accept product's key in products.
   * For example: onRemoveProduct('/shop/macbook-case/_red');
   * Required.
   *
   * @prop {React Element} CheckoutButton - Button in the bottom of cart.
   * Required.
   */
  containerPropTypes = {
    products: PropTypes.objectOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      quantity: PropTypes.number.isRequired,
      properties: PropTypes.object,
      productInfo: PropTypes.shape({
        name: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
        currency: PropTypes.string.isRequired,
        imagePath: PropTypes.string.isRequired,
      }).isRequired,
    })).isRequired,
    isCartEmpty: PropTypes.bool.isRequired,
    onUpdateProduct: PropTypes.func.isRequired,
    onRemoveProduct: PropTypes.func.isRequired,
    CheckoutButton: PropTypes.element.isRequired,
  },
  defaultProps = {
    productPropsToShow: ['colour', 'size'],
    showHeader: true,
    iconTrashClassName: 'icon-trash',
  };

/*
 * Create form-group for each of properties which is in productPropsToShow
 * array
 * Returning array may contain false values (when prop is in properties, but not visible)
 */
function createProductDescription(
  { properties } : Object,
  productPropsToShow : Array<string>
) : Array<React$Element<any> | false> {
  return Object.entries(properties).map(([k, v]) =>
  !!(productPropsToShow.indexOf(k) + 1) &&
  (
    <div key={ k + v } className="form-group row">
      <label htmlFor={ k } className="col-xs-6 col-md-5 col-lg-4 col-form-label" >
        { capitalize(k) }:
      </label>
      <div className="col-xs-6 col-md-7 col-lg-8 col-form-label">
        { capitalize(v) }
      </div>
    </div>)
  );
}

/*
 * Create product's list item
 */
function createProductGroup(
  products : ProductsMapType,
  productPropsToShow : Array<string>,
  onUpdateProduct : onUpdateProductType,
  onRemoveProduct : onRemoveProductType,
  iconTrashClassName : string) : Array<React$Element<any>> {
  return Object.entries(products).map(([productKey, { id, ...product, productInfo }]) => (
    <div key={ productKey } className="list-group-item list-group-item-action animated">
      <Link to={ id }>
        <div className="list-group-item-heading">
          <h5>{ productInfo.name }</h5>
        </div>
      </Link>
      <div className="list-group-item-text row">
        <div className="col-xs-12 col-sm-4 col-md-4 col-lg-4 m-b-1">
          <Link to={ id }>
            <img className="img-fluid" src={ productInfo.imagePath } />
          </Link>
        </div>
        <div className="col-xs-12 col-sm-4 col-md-4 col-lg-5">
          <div className="form-group row">
            <label htmlFor="quantity" className="col-xs-6 col-md-5 col-lg-4 col-form-label" >Quantity:</label>
            <div className="col-xs-6 col-md-7 col-lg-8">
              <input
                type="number"
                className="form-control"
                value={ product.quantity }
                onChange={
                  ({ target: { value } }) =>
                    /*
                     * Check if quantity value is correct and then update product
                     */
                    +value > -1 ?
                    onUpdateProduct(productKey, { quantity: +value })
                    : 0
                }
              />
            </div>

          </div>
          { createProductDescription(product, productPropsToShow) }
          <div className="form-group row">
            <label htmlFor="price" className="col-xs-6 col-md-5 col-lg-4 col-form-label" >
              Price:
            </label>
            <div className="col-xs-6 col-md-7 col-lg-8 col-form-label">
              { productInfo.currency + (+productInfo.price).toFixed(2) }
            </div>
          </div>
          <div className="form-group row">
            <label htmlFor="total" className="col-xs-6 col-md-5 col-lg-4 col-form-label" >
              Total:
            </label>
            <div className="col-xs-6 col-md-7 col-lg-8 col-form-label">
              { productInfo.currency + (productInfo.price * product.quantity).toFixed(2) }
            </div>
          </div>
        </div>
        <div className="col-xs-12 col-sm-4 col-md-4 col-lg-3">
          <div className="form-group row">
            <div className="col-md-12 text-xs-center">
              <button
                className="btn btn-danger form-control"
                onClick={
                  () => onRemoveProduct(productKey)
                }
              >
                <i className={ iconTrashClassName } />
                Remove
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    )
  );
}

function Cart (
  /*
   * Look at the propTypes
   */
  {
    productPropsToShow,
    showHeader,
    products,
    isCartEmpty,
    onUpdateProduct,
    onRemoveProduct,
    CheckoutButton,
    iconTrashClassName,
  } : Object) : React$Element<any> {
  return (
    <div className="row m-t-1">
      <Transition
        style={ animate(500) }
        in={ !isCartEmpty }
        enteringClassName="fadeInUp"
        exitingClassName="fadeOut"
        timeout={ 500 }
        unmountOnExit
      >
        <div className="col-xs-12 col-sm-12 col-md-12 col-lg-12">
          { showHeader ? <h4>Shopping cart</h4> : void 0 }
          <div className="list-group">
            <ReactCSSTransitionGroup
              transitionName={ {
                enter: 'bounceInLeft',
                leave: 'bounceOutRight',
              } }
              transitionEnterTimeout={ 500 } transitionLeaveTimeout={ 500 }
            >
              {
                createProductGroup(
                  products,
                  productPropsToShow,
                  onUpdateProduct,
                  onRemoveProduct,
                  iconTrashClassName
                )
              }
            </ReactCSSTransitionGroup>
          </div>
          <div className="row m-t-1">
            <div className="col-xs-0 col-sm-2 col-md-2 col-lg-3 col-xl-3" />
            <div className="col-xs-12 col-sm-8 col-md-8 col-lg-6 col-xl-6">
              { CheckoutButton }
            </div>
          </div>
        </div>
      </Transition>
    </div>
  );
}

Cart.propTypes = { ...propTypes, ...containerPropTypes };
Cart.defaultProps = defaultProps;

export default Cart;
