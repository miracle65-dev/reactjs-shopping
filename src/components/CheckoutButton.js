/**
 * @flow
 * @module CheckoutButton
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * React stateless component.
 * Checkout button with grand total.
 *
 */
import React, { Component, PropTypes } from 'react';
import Link from 'react-router/lib/Link';
import Transition from 'react-overlays/lib/Transition';

import { animate } from '../helpers';

const
/**
  * @static propTypes
  * @memberof CheckoutButton
  *
  * @prop {string} checkoutURL - Link to checkout page.
  * Default is '/shop/checkout/'
  * @prop {string} iconCheckoutClassName - ClassName for cart icon on checkout button.
  * Default is 'icon-basket'
  */
  propTypes = {
    checkoutURL: PropTypes.string,
    iconCheckoutClassName: PropTypes.string,
  },
/**
  * @static containerPropTypes
  * @memberof CheckoutButton
  *
  * @prop {number} grandTotal - Amount of money to pay. Required
  * @prop {boolean} hidden - Show or hide button. Required
  */
  containerPropTypes = {
    grandTotal: PropTypes.number.isRequired,
    hidden: PropTypes.bool.isRequired,
  },
  defaultProps = {
    checkoutURL: '/shop/checkout/',
    iconCheckoutClassName: 'icon-basket',
  };

function CheckoutButton (
  { grandTotal, hidden, checkoutURL, iconCheckoutClassName } : Object
) : React$Element<any> {
  return (
    <Transition
      style={ animate(500) }
      in={ !hidden }
      enteringClassName="fadeInUp"
      exitingClassName="fadeOut"
      timeout={ 500 }
      unmountOnExit
    >
      <Link to={ checkoutURL } className="btn btn-primary btn-block">
        <i className={ iconCheckoutClassName } />
        Checkout (Grand total Â£{ grandTotal.toFixed(2) })
      </Link>
    </Transition>
  );
}

CheckoutButton.propTypes = { ...propTypes, ...containerPropTypes };
CheckoutButton.defaultProps = defaultProps;

export default CheckoutButton;
