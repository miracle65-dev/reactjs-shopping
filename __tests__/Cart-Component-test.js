/*
 * Cart tests for JEST
 *
 * Copyright © Oleg Nosov 2016
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import IntlMessageFormat from 'intl-messageformat';
import { mount } from 'enzyme';
import { Cart } from '../src/components';

const testCartLocalization = {
  colour: 'Colour',
  iPadCase: 'iPad case',
  red: 'Red',
  shoppingCartTitle: 'My lovely shopping cart',
  productName: '{localizedName}',
  quantityLabel: 'Quantity:',
  priceLabel: 'Price:',
  priceValue: '{localizedCurrency}{price}',
  totalLabel: 'Total:',
  totalValue: '{localizedCurrency}{total, plural, ' +
              '=0 {0}' +
              'other {#}}',
  remove: 'Remove',
  productPropertyLabel: '{localizedName}:',
  productPropertyValue: '{localizedValue}',
  GBP: '£',
};

const getLocalization = (id, params = {}) =>
  new IntlMessageFormat(testCartLocalization[id]).format(params);

const createCart = ({ products }, renderFunc = mount) =>
  renderFunc(
    <Cart
      products={{ ...products }}
      onUpdateProduct={
        (key, updateProperties) =>
          // That's just a test. Don't do like this,
          // use immutable structures, please :C
          void (products[key] = { ...products[key], ...updateProperties })
      }
      onRemoveProduct={
        key => void delete products[key]
        // Don't mutate given objects, create new
      }
      CheckoutButton={
        <a />
      }
      isCartEmpty={
        false
      }
      getLocalization={
        getLocalization
      }
      currency="GBP"
    />,
  );


describe('Cart', () => {
  const iPadCaseInCart = {
    id: 'ipad-case',
    properties: {
      colour: 'red',
    },
    quantity: 1,
    productInfo: {
      name: 'iPadCase',
      prices: {
        GBP: 70,
      },
      imagePath: 'ipad-case1-483x321.jpeg',
      propertiesToShowInCart: ['colour'],
      path: '/shop/ipad-case',
    },
  };

  it('changes product quantity value', () => {
    // Create a cart with 1 red iPad case in
    const products = { 'ipad-case_red': iPadCaseInCart };

    const renderedCart = createCart({ products });

    // Change iPad cases quantity to 10
    renderedCart.find('input').simulate('change', { target: { value: 10 } });

    // Now must have 10 cases in cart
    expect(products['ipad-case_red'].quantity).toBe(10);

    // Change iPad cases quantity to -1
    renderedCart.find('input').simulate('change', { target: { value: -1 } });

    // Expect 10 iPad cases in cart
    expect(products['ipad-case_red'].quantity).toBe(10);

    // Remove our cases from cart
    renderedCart.find('button').simulate('click');

    // Now cart must be empty
    expect(Object.keys(products).length).toBe(0);
  });

  it('takes snapshot', () => {
    const products = { 'ipad-case_red': iPadCaseInCart };
    const renderedCart = createCart({ products }, renderer.create);
    expect(renderedCart.toJSON()).toMatchSnapshot();
  });
});
