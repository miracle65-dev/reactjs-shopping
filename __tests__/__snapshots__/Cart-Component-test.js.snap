exports[`Cart takes snapshot 1`] = `
<div
  className="row mt-1">
  <div
    className="col-12"
    style={
      Object {
        "WebkitAnimationDuration": "0.5s",
        "WebkitAnimationFillMode": "both",
        "animationDuration": "0.5s",
        "animationFillMode": "both"
      }
    }>
    My lovely shopping cart
    <div
      className="list-group">
      <span>
        <div
          className="list-group-item list-group-item-action align-items-start animated">
          <div
            className="col-xs-12 col-sm-12 col-md-4 col-lg-4 col-xl-3">
            <a
              href="/shop/ipad-case"
              to="/shop/ipad-case">
              <div
                className="list-group-item-heading">
                iPad case
              </div>
              <img
                className="img-fluid"
                src="ipad-case1-483x321.jpeg" />
            </a>
          </div>
          <div
            className="col-xs-12 col-sm-12 col-md-5 col-lg-5 col-xl-7">
            <div
              className="form-group row">
              <label
                className="col-xs-6 col-md-5 col-lg-4 col-form-label"
                htmlFor="quantity">
                Quantity:
              </label>
              <div
                className="col-xs-6 col-md-7 col-lg-8">
                <input
                  className="form-control"
                  onChange={[Function anonymous]}
                  type="number"
                  value={1} />
              </div>
            </div>
            <div
              className="form-group row">
              <label
                className="col-xs-6 col-md-5 col-lg-4 col-form-label"
                htmlFor="colour">
                Colour:
              </label>
              <div
                className="col-xs-6 col-md-7 col-lg-8 col-form-label">
                Red
              </div>
            </div>
            <div
              className="form-group row">
              <label
                className="col-xs-6 col-md-5 col-lg-4 col-form-label"
                htmlFor="price">
                Price:
              </label>
              <div
                className="col-xs-6 col-md-7 col-lg-8 col-form-label">
                £70
              </div>
            </div>
            <div
              className="form-group row">
              <label
                className="col-xs-6 col-md-5 col-lg-4 col-form-label"
                htmlFor="total">
                Total:
              </label>
              <div
                className="col-xs-6 col-md-7 col-lg-8 col-form-label">
                £70
              </div>
            </div>
          </div>
          <div
            className="col-xs-12 col-sm-12 col-md-3 col-lg-3 col-xl-2">
            <div
              className="form-group row">
              <div
                className="col-12 text-center">
                <button
                  className="btn btn-danger active form-control"
                  onClick={[Function anonymous]}
                  role="button">
                  <i
                    className="icon-trash" />
                  Remove
                </button>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
    <div
      className="row mt-1">
      <div
        className="col-xs-12 col-sm-8 col-md-8 col-lg-6 col-xl-6offset-xs-0 offset-sm-2 offset-md-2 offset-lg-3 offset-xl-3">
        <a />
      </div>
    </div>
  </div>
</div>
`;
